locale.c

```diff
@@ -8225,34 +8253,84 @@ S_ints_to_tm(pTHX_ struct tm * mytm,
     mytm->tm_mday = mday;
     mytm->tm_mon = mon;
     mytm->tm_year = year;
-    mytm->tm_wday = wday;
-    mytm->tm_yday = yday;
-    mytm->tm_isdst = isdst;
 
-    /* Long-standing behavior is to ignore the effects of locale (in
-     * particular, daylight savings time) on the input, so we use mini_mktime.
-     * See GH #22062. */
+    struct tm * which_tm = mytm;
+    struct tm aux_tm;
+
+#ifndef HAS_MKTIME
+
     mini_mktime(mytm);
 
-    /* But some of those effect are deemed desirable, so use libc to get the
-     * values for tm_gmtoff and tm_zone on platforms that have them [perl
-     * #18238] */
-#if  defined(HAS_MKTIME)                                      \
- && (defined(HAS_TM_TM_GMTOFF) || defined(HAS_TM_TM_ZONE))
+#else
 
-    const char * orig_TIME_locale = toggle_locale_c(LC_TIME, locale);
-    struct tm aux_tm = *mytm;
-    MKTIME_LOCK;
-    mktime(&aux_tm);
-    MKTIME_UNLOCK;
-    restore_toggled_locale_c(LC_TIME, orig_TIME_locale);
+    /* On platforms that have either of these two fields, we have to run the
+     * libc mktime() in order to set them, as mini_mktime() doesn't deal with
+     * them.  [perl #18238] */
+#  if defined(HAS_TM_TM_GMTOFF) || defined(HAS_TM_TM_ZONE)
+#    define ALWAYS_RUN_MKTIME
+#  endif
+
+    /* When isdst is 0, it means to consider daylight savings time as never
+     * being in effect.  Many libc implementations of mktime() do not allow
+     * this; they always consider the possibility of dst.  But mini_mktime()
+     * never considers dst, so use it under this condition. */
+    if (isdst == 0) {
+        mini_mktime(mytm);
+
+#  ifndef ALWAYS_RUN_MKTIME
+
+    /* When we don't always need libc mktime(), we call it only when we didn't
+     * call mini_mktime() */
+    } else {
+
+#  else
+        /* Here will have to run libc mktime() in order to get the values of
+         * some fields that mini_mktime doesn't populate.  We don't want
+         * mktime's side effect of looking for dst, so we have to have a
+         * separate tm structure from which we copy just those fields into the
+         * returned structure.  Initialize its values.  mytm should now be a
+         * normalized version of the input. */
+        StructCopy(mytm, &aux_tm, struct tm);
```

Comment:
Why use StructCopy() here instead of simple assignment?

StructCopy() ignores the type and const-ness of the source and target, which isn't needed here.

Reply:
It was defensive coding because I mistakenly thought StructCopy did a deep copy, and if `struct tm` were ever changed to have pointers, this would still work.  But the odds of that were close to zero anyway.

Reply:
From looking at the history, it appears to be there to support pre-ANSI C compilers, if the compiler didn't support structure assignment it did a Copy() instead.